src/test/java/test/beforegroups/BeforeGroupsTest.java

```diff
@@ -0,0 +1,65 @@
+package test.beforegroups;
+
+import org.testng.ITestNGListener;
+import org.testng.TestNG;
+import org.testng.annotations.Test;
+import org.testng.collections.Lists;
+import org.testng.internal.ClassHelper;
+import org.testng.internal.PackageUtils;
+import org.testng.xml.XmlSuite;
+import org.testng.xml.XmlTest;
+import test.InvokedMethodNameListener;
+import test.SimpleBaseTest;
+import test.beforegroups.issue1694.BaseClassWithBeforeGroups;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class BeforeGroupsTest extends SimpleBaseTest {
+    @Test
+    public void testInSequentialMode() throws IOException {
+        runTest(XmlSuite.ParallelMode.NONE);
+    }
+
+    @Test
+    public void testParallelMode() throws IOException {
+        runTest(XmlSuite.ParallelMode.CLASSES);
+    }
+
+    private static void runTest(XmlSuite.ParallelMode mode) throws IOException {
+        XmlSuite suite = createXmlSuite("sample_suite");
+        String pkg = BaseClassWithBeforeGroups.class.getPackage().getName();
+        Class<?>[] classes = findClassesInPackage(pkg);
+        XmlTest xmlTest = createXmlTestWithPackages(suite, "sample_test", classes);
+        xmlTest.addIncludedGroup("regression");
+        xmlTest.setParallel(mode);
+        TestNG tng = create(suite);
+        InvokedMethodNameListener listener = new InvokedMethodNameListener();
+        tng.addListener((ITestNGListener) listener);
+        tng.run();
+        List<String> beforeGroups = Lists.newArrayList();
+        List<String> afterGroups = Lists.newArrayList();
+        for (String name : listener.getInvokedMethodNames()) {
+            if (name.equalsIgnoreCase(BaseClassWithBeforeGroups.BEFORE_GROUPS)) {
+                beforeGroups.add(name);
+            }
+            if (name.equalsIgnoreCase(BaseClassWithBeforeGroups.AFTER_GROUPS)) {
+                afterGroups.add(name);
+            }
+        }
+        assertThat(beforeGroups).containsOnly(BaseClassWithBeforeGroups.BEFORE_GROUPS);
```

Comment:
You are filtering methods before, so the assertion will be always true :)
The test should fail if there is more invoked methods than the ones we expect.

Reply:
Not necessarily true. The filtering is being done after the execution, and we are filtering only the `beforeGroups()` and `afterGroups()` methods so that assertion is easy. I have double-checked this to ensure that is the case.

Reply:
I don't catch it:
```
if (name.equalsIgnoreCase(BaseClassWithBeforeGroups.BEFORE_GROUPS)) {
  beforeGroups.add(name);
}
```
enforces than beforeGroups only contains BEFORE_GROUPS values.
So, `assertThat(beforeGroups).containsOnly(BaseClassWithBeforeGroups.BEFORE_GROUPS);` is always true.

At the same time, what about the other invoked methods? And what if we have many or none BEFORE_GROUPS (if I remember well, containsOnly doesn't care about an empty collection or duplicate values)?

Reply:
@juherr 

> enforces than beforeGroups only contains BEFORE_GROUPS values.

`BEFORE_GROUPS` is a constant. 

The below sample should clarify your question about `assertion being true always`

```java
List<String> strings = Arrays.asList("testng", "junit");
Assertions.assertThat(strings).containsExactly("testng");
```

output
```
Exception in thread "main" java.lang.AssertionError: 
Actual and expected should have same size but actual size was:
  <2>
while expected size was:
  <1>
Actual was:
  <["testng", "junit"]>
Expected was:
  <["testng"]>
```

> At the same time, what about the other invoked methods? 

The test is to ensure that beforeGroups() gets called exactly once for the group. The changes are not related to other methods. So I haven't run any assertions for them.

> And what if we have many or none BEFORE_GROUPS 

If there are no `@BeforeGroups` method, then its existing functionality I guess? That is why I didn't have any tests for that.

(if I remember well, containsOnly doesn't care about an empty collection or duplicate values)?

Please let me know if that answers your question.

Reply:
Yes, but what I am reading is:

```java
List<String> strings = Arrays.asList("testng", "junit");
List<String> assertedStrings = ...;
for (String s : strings) {
   if (s.equals("testng) {
    assertedStrings.add(s);
  }
}
Assertions.assertThat(assertedStrings).containsOnly("testng");
```

Reply:
Yes that is the intent here, because I want to filter out all occurrences of `testng` and then ensure that the occurrence is exactly 1.